src/systems/filecoin_vm/runtime/runtime.go

```diff
@@ -1,81 +1,424 @@
+	rt._returnInternal(msg.InvocOutput_Make(exitCode, nil))
+	return &Runtime_ReturnError_FunRet_I{}
+}
+
+func (rt *Runtime_I) _throwError(exitCode ExitCode) {
+	rt._throwErrorFull(exitCode, "")
+}
+
+func (rt *Runtime_I) _throwErrorFull(exitCode ExitCode, errMsg string) {
+	panic(exitcode.RuntimeError_Make(exitCode, errMsg))
+}
+
+func (rt *Runtime_I) _apiError(errMsg string) {
+	rt._throwErrorFull(exitcode.SystemError(exitcode.RuntimeAPIError), errMsg)
+}
+
+func (rt *Runtime_I) _checkStateLock(expected bool) {
+	if rt._actorStateAcquired() != expected {
+		rt._apiError("State update and message send blocks must be disjoint")
+	}
+}
+
+func (rt *Runtime_I) _checkGasRemaining() {
+	if rt._gasRemaining().LessThan(msg.GasAmount_Zero()) {
+		rt._throwError(exitcode.SystemError(exitcode.OutOfGas))
+	}
+}
+
+func (rt *Runtime_I) _deductGasRemaining(x msg.GasAmount) {
+	// TODO: check x >= 0
+	rt._checkGasRemaining()
+	rt._gasRemaining_ = rt._gasRemaining().Subtract(x)
+	rt._checkGasRemaining()
+}
+
+func (rt *Runtime_I) _refundGasRemaining(x msg.GasAmount) {
+	// TODO: check x >= 0
+	rt._checkGasRemaining()
+	rt._gasRemaining_ = rt._gasRemaining().Add(x)
+	rt._checkGasRemaining()
+}
+
+func (rt *Runtime_I) _transferFunds(from addr.Address, to addr.Address, amount actor.TokenAmount) {
+	rt._checkRunning()
+	rt._checkStateLock(false)
+
+	transferRet := rt._globalStatePending().WithFundsTransfer(from, to, amount)
+	if transferRet.Which() == st.StateTree_WithFundsTransfer_FunRet_Case_error {
+		rt._throwError(exitcode.SystemError(exitcode.InsufficientFunds))
+	} else {
+		rt._globalStatePending_ = transferRet.As_StateTree()
+	}
+}
+
+func (rt *Runtime_I) _incrementCallSeqNum(a addr.Address) {
 	panic("TODO")
 }
 
-func (s *VMState_I) ComputeActorAddress(creator addr.Address, nonce actor.CallSeqNum) addr.Address {
+// TODO: This function should be private (not intended to be exposed to actors).
+// (merging runtime and interpreter packages should solve this)
+func (rt *Runtime_I) SendToplevelFromInterpreter(to addr.Address, input InvocInput, ignoreErrors bool) (
+	msg.MessageReceipt, st.StateTree) {
+
+	rt._running_ = true
+	ret := rt._sendInternal(to, input, ignoreErrors)
+	rt._running_ = false
+	return ret, rt._globalStatePending()
+}
+
+func _invokeMethodInternal(
+	rt *Runtime_I,
+	actorCode ActorCode,
+	method actor.MethodNum,
+	params actor.MethodParams) (ret InvocOutput, gasUsed msg.GasAmount) {
+
+	if method == Reserved_NoopMethod {
+		ret = msg.InvocOutput_Make(exitcode.OK(), nil)
+		gasUsed = msg.GasAmount_Zero() // TODO: verify
+		return
+	}
+
+	rt._running_ = true
+	catchRetCode := exitcode.CatchRuntimeErrors(func() {
+		actorCode.InvokeMethod(rt, method, params)
+		rt._checkStateLock(false)
+		rt._checkNumValidateCalls(1)
+		rt._checkNumReturnCalls(1)
+		for _, item := range rt._messageQueue() {
+			rt._sendInternal(item.to, item.input, item.ignoreErrors)
+		}
+		rt._messageQueue_ = []MessageQueueItem{}
```

Comment:
this got significantly more complicated, do we actually need this? (eg the message queue)

Reply:
(removed deferred send for now as per discussion)